### PR TITLE
gRPC server streaming

### PR DESCRIPTION
**Pull request by jumaffre**
Addresses #4325 

This is now ready for review. 

This PR implements server-side gRPC streaming, either by streaming synchronously from within the endpoint handler, or by detaching the stream and streaming asynchronously, e.g. from within the indexer.

Notable changes:
- HTTP/2 support for server-side streaming, by returning `NGHTTP2_ERR_DEFERRED` from within `read_outgoing_callback`.
- New `ccf::grpc::Stream` and `ccf::grpc::DetachedStream` objects accessible from within the application via new `grpc_command_unary_stream_adapter` endpoint wrapper.

Still to be done (separate PR):
- Provide close callback to `DetachedStream` so that app can perform cleanup when the client session is closed asynchronously.

TODO:
- [x] API: Create separate detachable stream that can be closed (!= inline streaming)
- [x] Streaming endpoint should return nothing (i.e. `make_success()`). 
- [x] Fix for when synchronous stream returns no message
- [x] Pub-sub endpoint
- [x] Attach `lookup_responder` to `RpcContext`. 
- [x] HTTP/2 layer cleanup
- [x] Fix segfault when detached stream responds when client has closed stream
- [x] Split stream close
- [x] Support for no data streamed in inline streaming endpoint

### CODE REVIEW

File: src/apps/external_executor/external_executor.cpp
Comment (eddyashton) [inline src/apps/external_executor/external_executor.cpp:568]: Passing a `close_stream` bool to each call seems cumbersome. Preferably we'd just have a final `stream->close_stream()` call, and maybe even call that implicitly from the `Stream`'s destructor if we're feeling RAIIdical.

File: src/apps/external_executor/external_executor.cpp
Comment (eddyashton) [inline src/apps/external_executor/external_executor.cpp:573]: I think something like `return ccf::grpc::pending();` would actually be clearer than `ccf::grpc::make_success();`. But it may also be better to avoid these return-based wrappers for streaming endpoints? So `grpc_adapter` gives us 2 features; type-safety, and a transformation to stack-constrained API (here's your params, return your result). We want to keep the former, but the latter can be dropped for these if it makes sense.

File: src/apps/external_executor/external_executor.cpp
Comment (eddyashton) [inline src/apps/external_executor/external_executor.cpp:188]: I think we should add a pub/sub endpoint (ie - one streaming endpoint which gets another message contains every time `/foo` is called) as a sample, in this PR. The underlying implementation is also tested by this endpoint, but that would prove out the `Stream` ownership model/reachability a little more (this sample still triggers everything on the `Stream` from a single user call - albeit transitively via the delayed thread message - and we want to confirm we can also write responses triggered from other endpoints, or hooks, or indexes, or...)

**Review by eddyashton — COMMENTED** [review]
Looks great. Have added some very early comments on the API, let's discuss further tomorrow.

File: include/ccf/rpc_context.h
Comment (MahatiC) [reply include/ccf/rpc_context.h:147]: I chatted with @jumaffre and it will be eventually supported

File: src/apps/external_executor/external_executor.cpp
Comment (jumaffre) [reply src/apps/external_executor/external_executor.cpp:179]: I've added a `/Pub` and `/Sub` endpoints. I'll leave the `/Stream` endpoint in this PR until just before this PR is merged just to demonstrate multiple examples.

File: src/apps/external_executor/external_executor.cpp
Comment (jumaffre) [reply src/apps/external_executor/external_executor.cpp]: Done. There's now a separate `close()` function, called from within the destructor too.

File: src/apps/external_executor/external_executor.cpp
Comment (jumaffre) [reply src/apps/external_executor/external_executor.cpp]: Done. The adapter is now much cleaner, and creates the `Stream` for the handler.

File: src/apps/external_executor/external_executor.cpp
Comment (jumaffre) [inline src/apps/external_executor/external_executor.cpp:177]: Remove async task endpoint before merging this PR.

File: include/ccf/rpc_context.h
Comment (eddyashton) [inline include/ccf/rpc_context.h:146]: Do we need this overload? We don't have it for `set_response_header`, and it seems weird to make those overload sets different.

File: src/apps/external_executor/external_executor.cpp
Comment (eddyashton) [inline src/apps/external_executor/external_executor.cpp:329]: IUIC, this is now constructing a `std::vector` here, rather than a span pointing at the existing payload bytes? That's OK for now because the `std::vector` is moved directly into a buffer in the implementation (so there would already have been a copy there, we've just moved it here). But if we're able to copy directly to the target memory range in this `send_response` call, then we'd prefer to copy _from_ the original byes (with a `span`) than this intermediate `vector`.

I think we should prefer spans in general, and only do copies to an owning `std::vector` internally/as late as possible. That makes it possible to eventually remove intermediate copies entirely, whereas `std::vector`s in the API suggest a requirement for these intermediate copies.

File: src/apps/external_executor/external_executor.cpp
Comment (eddyashton) [inline src/apps/external_executor/external_executor.cpp:783]: This sample isn't actually interacting with the KV at all, so its odd that it uses `KVKeyValue`. I think it's a dangerous sample, because it _looks_ like a event-based notification of KV modifications, but this is executing entirely before commit/consensus.

I think these Pub/Sub endpoints should use new message types, with a simpler key indexing into the `subscribed_keys` table. If we want to build a sample of pub-sub event notification of changes to the KV, I think we need to do that in an indexer (and think about/comment on the reliability guarantees).

File: src/endpoints/grpc/grpc.h
Comment (eddyashton) [inline src/endpoints/grpc/grpc.h:42]: I don't think that's safe - what if the endpoint wants to overwrite these default headers? I think this is saying they might be able to, if they call `set_response_header` and overwrite this _before_ their endpoint calls something which triggers a data stream send? If that's the API, we need to make those send calls extremely obvious, and/or make header modification _after_ that send return an error.

File: tests/external_executor/external_executor.py
Comment (MahatiC) [inline tests/external_executor/external_executor.py:471]: why do we need this timeout?

File: src/endpoints/grpc/grpc.h
Comment (jumaffre) [reply src/endpoints/grpc/grpc.h:42]: Good spot. This was only moved there before I added the `ccf::grpc::Stream` object so I've moved this back to `set_grpc_response`.

File: tests/external_executor/external_executor.py
Comment (jumaffre) [reply tests/external_executor/external_executor.py:542]: Merge conflict gone wrong it seems. I have reverted.

File: tests/external_executor/external_executor.py
Comment (jumaffre) [reply tests/external_executor/external_executor.py:471]: I've removed the timeout.

File: include/ccf/rpc_context.h
Comment (jumaffre) [reply include/ccf/rpc_context.h:146]: I added it so that we can call `ctx->set_response_trailer(make_status_trailer(response.status.code()));` as it has to be called in two different places (unary and stream trailers), and I'd forgotten to call `std::to_string(response.status.code())` in one of them, which compiled fine, but caused weird errors at runtime. I'm open to another way of doing this.

File: src/apps/external_executor/external_executor.cpp
Comment (jumaffre) [reply src/apps/external_executor/external_executor.cpp]: Done.

File: src/apps/external_executor/external_executor.cpp
Comment (jumaffre) [reply src/apps/external_executor/external_executor.cpp]: Good point. I've changed this back to a `std::span`, which is now also used for streaming data too.

File: src/apps/external_executor/protobuf/misc.proto
Comment (MahatiC) [inline src/apps/external_executor/protobuf/misc.proto:5]: If this is intended to be an example, maybe we can rename this to `example`

Comment (jumaffre) [reply src/apps/external_executor/protobuf/misc.proto:5]: Yes. We will move the gRPC-specific endpoints to a new app altogether but I'll leave this to a further PR.

File: src/apps/external_executor/external_executor.cpp
Comment (achamayou) [inline src/apps/external_executor/external_executor.cpp:744]: Does that need a ticket?

File: src/http/http2_types.h
Comment (achamayou) [inline src/http/http2_types.h:39]: Not sure why we have the span here? That looks like the ownership is with data, we can return a fresh span when necessary?

**Review by achamayou — APPROVED** [review]

File: src/apps/external_executor/external_executor.cpp
Comment (jumaffre) [reply src/apps/external_executor/external_executor.cpp:744]: This is item 5. in https://github.com/microsoft/CCF/issues/4325.

File: src/http/http2_utils.h
Comment (jumaffre) [reply src/http/http2_utils.h:16]: This is item 8. in https://github.com/microsoft/CCF/issues/3342.

File: src/http/http_rpc_context.h
Comment (jumaffre) [reply src/http/http_rpc_context.h:21]: Yes. I think this work is part of the wider "avoid copies whenever possible" in the HTTP/2 stack (see https://github.com/microsoft/CCF/issues/3342).

File: src/http/http2_types.h
Comment (eddyashton) [reply src/http/http2_types.h:39]: For posterity:

This struct is used to store data we're trying to `send` through http2. When http2 calls us back to copy this data, its target buffer has a capped size. So if we're trying to send a 10MB `data`, we may get called multiple times and told to copy in just 4K at a time. From our perspective, this all happens under a single call we made, but http2 insists on chunking it per-frame or something. When that happens, instead of truncating the start of `data` each time (with potential reallocation/realignment costs) to indicate how much we've written, or maintaining a separate `offset` counting how many bytes we've written, we thought it was neater to use a `span` - it always points at a subrange (and in fact suffix) of `data`, but is cheap to trim the start off of as we advance.

Comment (achamayou) [reply src/http/http2_types.h:39]: @jumaffre can we put what @eddyashton wrote in the comment above? :)

**Pull request merged**
